### PR TITLE
Fix memory leaks in CA related functions.

### PR DESCRIPTION
**Pull request by PavelKopyl**
Fix memory leaks in CA related functions.
CLA: trivial

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by mattcaswell** [comment]
@PavelKopyl do you think you could submit a CLA please? I think we are starting to stray beyond the "trivial" exception.

**Comment by PavelKopyl** [comment]
Ok, I'll start an approval to sign the CLA, but it may take some time.

**Comment by mattcaswell** [comment]
If this is a corporate contribution then you will need to send both a CCLA from your company and an ICLA for yourself.

File: apps/ca.c [hunk 1]
```diff
@@ -940,8 +940,10 @@ int ca_main(int argc, char **argv)
             if (j > 0) {
                 total_done++;
                 BIO_printf(bio_err, "\n");
-                if (!BN_add_word(serial, 1))
+                if (!BN_add_word(serial, 1)) {
+                    X509_free(x);
                     goto end;
+                }
                 if (!sk_X509_push(cert_sk, x)) {
                     BIO_printf(bio_err, "Memory allocation failure\n");
```
Comment (mattcaswell) [inline apps/ca.c]: Shouldn't we also free `x` here too?

Comment (PavelKopyl) [reply apps/ca.c]: Yes, you're right.

Comment (PavelKopyl) [reply apps/ca.c]: Some changes are required to apply it to 1.1.0 and 1.0.2. I'll make appropriate pull requests.

File: crypto/engine/eng_lib.c [hunk 2]
```diff
@@ -149,8 +149,10 @@ void engine_cleanup_add_last(ENGINE_CLEANUP_CB *cb)
     if (!int_cleanup_check(1))
         return;
     item = int_cleanup_item(cb);
-    if (item)
-        sk_ENGINE_CLEANUP_ITEM_push(cleanup_stack, item);
+    if (item) {
```
Comment (richsalz) [inline crypto/engine/eng_lib.c]: Should that be tested against NULL?  Not worth delaying this PR if it will take awhile to get around to fixing that nit.

**Comment by mattcaswell** [comment]
Looks like this is in need of a rebase.

**Comment by levitte** [comment]
The conflict in `apps/ca.c` is this:

```
<<<<<<< HEAD
    if (ok != 1) {
=======
    if (irow != NULL || ok <= 0) {
>>>>>>> Fix memory leaks in CA related functions.
```

As far as I can see, we can simply drop the `apps/ca.c` change and use the rest.  If a merge can be approved under those conditions, I can do it.

**Comment by levitte** [comment]
Merged into master / 1.1.1, with the test against NULL nit fixed.

aebd0e5ca12d1ba0b229a4121a54afa5ea2d8aa1 Fix memory leaks in CA related functions.

**Comment by levitte** [comment]
This doesn't cherry-pick cleanly to 1.1.0, although that's only because of a small nit, 1.1.0 has an assertion that isn't in 1.1.1:

```
diff --cc crypto/conf/conf_api.c
index 5535416ab3,9606b7f222..0000000000
--- a/crypto/conf/conf_api.c
+++ b/crypto/conf/conf_api.c
@@@ -204,7 -204,8 +204,12 @@@ CONF_VALUE *_CONF_new_section(CONF *con
      v->value = (char *)sk;
  
      vv = lh_CONF_VALUE_insert(conf->data, v);
++<<<<<<< HEAD
 +    OPENSSL_assert(vv == NULL);
++=======
+     if (vv != NULL || lh_CONF_VALUE_error(conf->data) > 0)
+         goto err;
++>>>>>>> aebd0e5ca1... Fix memory leaks in CA related functions.
      return v;
  
   err:
```

Porting this to 1.0.2 is more extensive.

**Comment by levitte** [comment]
See #6237 and #6238

**Comment by levitte** [comment]
The backports are merged, closing this.

**Pull request closed without merging**